File: mobile/lib/core/models/caderneta.dart
class Caderneta {
  final String id;
  final String childId;
  final String? childName;
  final String teacherId;
  final String? teacherName;
  final DateTime reportDate;
  final String? breakfastRating;
  final String? lunchRating;
  final String? snackRating;
  final String? physiologicalNeeds;
  final bool? hadNap;
  final String? sensorialMotorDevelopment;
  final String? intellectualDevelopment;
  final String? socialDevelopment;
  final String? affectiveDevelopment;
  final String? generalObservations;
  final String? behavior;
  final String? activities;
  final String? healthObservations;

  const Caderneta({
    required this.id,
    required this.childId,
    this.childName,
    required this.teacherId,
    this.teacherName,
    required this.reportDate,
    this.breakfastRating,
    this.lunchRating,
    this.snackRating,
    this.physiologicalNeeds,
    this.hadNap,
    this.sensorialMotorDevelopment,
    this.intellectualDevelopment,
    this.socialDevelopment,
    this.affectiveDevelopment,
    this.generalObservations,
    this.behavior,
    this.activities,
    this.healthObservations,
  });

  factory Caderneta.fromJson(Map<String, dynamic> json) {
    return Caderneta(
      id: json['id']?.toString() ?? '',
      childId: json['child_id']?.toString() ?? '',
      childName: () {
        final direct = json['child_name'] as String?;
        if (direct != null && direct.isNotEmpty) return direct;
        final composed = [json['first_name']?.toString(), json['last_name']?.toString()]
            .where((s) => s != null && s.isNotEmpty)
            .join(' ');
        return composed.isNotEmpty ? composed : null;
      }(),
      teacherId: json['teacher_id']?.toString() ?? '',
      teacherName: json['teacher_name'] as String?,
      reportDate: json['report_date'] != null
          ? DateTime.tryParse(json['report_date'] as String) ?? DateTime.now()
          : DateTime.now(),
      breakfastRating: json['breakfast_rating'] as String?,
      lunchRating: json['lunch_rating'] as String?,
      snackRating: json['snack_rating'] as String?,
      physiologicalNeeds: json['physiological_needs'] as String?,
      hadNap: json['had_nap'] as bool?,
      sensorialMotorDevelopment: json['sensorial_motor_development'] as String?,
      intellectualDevelopment: json['intellectual_development'] as String?,
      socialDevelopment: json['social_development'] as String?,
      affectiveDevelopment: json['affective_development'] as String?,
      generalObservations: json['general_observations'] as String?,
      behavior: json['behavior'] as String?,
      activities: json['activities'] as String?,
      healthObservations: json['health_observations'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'child_id': childId,
        'teacher_id': teacherId,
        'report_date':
            '${reportDate.year.toString().padLeft(4, '0')}-${reportDate.month.toString().padLeft(2, '0')}-${reportDate.day.toString().padLeft(2, '0')}',
        if (breakfastRating != null) 'breakfast_rating': breakfastRating,
        if (lunchRating != null) 'lunch_rating': lunchRating,
        if (snackRating != null) 'snack_rating': snackRating,
        if (physiologicalNeeds != null) 'physiological_needs': physiologicalNeeds,
        if (hadNap != null) 'had_nap': hadNap,
        if (sensorialMotorDevelopment != null)
          'sensorial_motor_development': sensorialMotorDevelopment,
        if (intellectualDevelopment != null)
          'intellectual_development': intellectualDevelopment,
        if (socialDevelopment != null) 'social_development': socialDevelopment,
        if (affectiveDevelopment != null)
          'affective_development': affectiveDevelopment,
        if (generalObservations != null)
          'general_observations': generalObservations,
        if (behavior != null) 'behavior': behavior,
        if (activities != null) 'activities': activities,
        if (healthObservations != null) 'health_observations': healthObservations,
      };
}

File: app/models/caderneta.py
import uuid
from datetime import date, datetime
from typing import Optional

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Index, String, Text, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class Caderneta(Base):
    __tablename__ = "cadernetas"
    __table_args__ = (
        UniqueConstraint("school_id", "child_id", "report_date", name="uq_caderneta_child_date"),
        Index("ix_cadernetas_school_id", "school_id"),
        Index("ix_cadernetas_child_id", "child_id"),
        Index("ix_cadernetas_report_date", "report_date"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    school_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schools.id", ondelete="RESTRICT"), nullable=False
    )
    child_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("children.id", ondelete="RESTRICT"), nullable=False
    )
    teacher_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("employees.id", ondelete="RESTRICT"), nullable=False
    )
    report_date: Mapped[date] = mapped_column(Date, default=date.today, nullable=False)
    breakfast_rating: Mapped[Optional[str]] = mapped_column(String(50))  # Bem, Muito Bem, Mal, Não Comeu
    lunch_rating: Mapped[Optional[str]] = mapped_column(String(50))
    snack_rating: Mapped[Optional[str]] = mapped_column(String(50))
    physiological_needs: Mapped[Optional[str]] = mapped_column(String(50))  # Normal, Mole, Duro
    had_nap: Mapped[Optional[bool]] = mapped_column(Boolean)
    sensorial_motor_development: Mapped[Optional[str]] = mapped_column(String(255))
    intellectual_development: Mapped[Optional[str]] = mapped_column(String(255))
    social_development: Mapped[Optional[str]] = mapped_column(String(255))
    affective_development: Mapped[Optional[str]] = mapped_column(String(255))
    general_observations: Mapped[Optional[str]] = mapped_column(Text)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
